spec.html

```diff
@@ -34020,17 +33899,31 @@ <h1>length</h1>
       <h1>Array Iterator Objects</h1>
       <p>An Array Iterator is an object, that represents a specific iteration over some specific Array instance object. There is not a named constructor for Array Iterator objects. Instead, Array iterator objects are created by calling certain methods of Array instance objects.</p>
 
-      <emu-clause id="sec-createarrayiterator" aoid="CreateArrayIterator">
+      <emu-clause id="sec-createarrayiterator" aoid="CreateArrayIterator" oldids="sec-properties-of-array-iterator-instances,table-48,table-internal-slots-of-array-iterator-instances">
         <h1>CreateArrayIterator ( _array_, _kind_ )</h1>
         <p>The abstract operation CreateArrayIterator takes arguments _array_ and _kind_. This operation is used to create iterator objects for Array methods that return such iterators. It performs the following steps when called:</p>
         <emu-alg>
           1. Assert: Type(_array_) is Object.
           1. Assert: _kind_ is ~key+value~, ~key~, or ~value~.
-          1. Let _iterator_ be ! OrdinaryObjectCreate(%ArrayIteratorPrototype%, &laquo; [[IteratedArrayLike]], [[ArrayLikeNextIndex]], [[ArrayLikeIterationKind]] &raquo;).
-          1. Set _iterator_.[[IteratedArrayLike]] to _array_.
-          1. Set _iterator_.[[ArrayLikeNextIndex]] to 0.
-          1. Set _iterator_.[[ArrayLikeIterationKind]] to _kind_.
-          1. Return _iterator_.
+          1. Let _closure_ be a new Abstract Closure with no parameters that captures _kind_ and _array_ and performs the following steps when called:
+            1. Let _index_ be 0.
+            1. Repeat,
+              1. If _array_ has a [[TypedArrayName]] internal slot, then
+                1. If IsDetachedBuffer(_array_.[[ViewedArrayBuffer]]) is *true*, throw a *TypeError* exception.
+                1. Let _len_ be _array_.[[ArrayLength]].
```

Comment:
The length of a `TypedArray` can never change after creation, so the step:

> 2. Let <var>len</var> be <var>array</var>.[[ArrayLength]].

can be extracted before the loop.

Reply:
That's true, but by convention (see https://github.com/tc39/ecma262/pull/2207) the [[ArrayLength]] field is only read after confirming that the underlying buffer is not detached, i.e., by a preceding `IsDetachedBuffer` check. That check is observable (in that we throw if it fails), and it needs to happen within the loop. So an editorial refactoring of this algorithm to hoist this step is probably not going to be an improvement in clarity. I think it's fine as-is.